### ISSUE TITLE
Update to Protobuf 4.26.0

### ISSUE THREAD

**Issue by neeme-praks-sympower** [grpc/grpc-java#11015]
### Is your feature request related to a problem?

I would like to use the latest released version of Protobuf (4.26.0) but I cannot as gRPC is still using the old version.

### Describe the solution you'd like

Update Protobuf dependency and make it work with the new version.

### Describe alternatives you've considered

I hoped that the update would be easy (just a version bump) so I tried doing it myself but it resulted in a compilation failure so I'll leave it to you to figure it out.

The failure:

```
> Task :grpc-protobuf:compileJava FAILED
grpc-java/protobuf/src/main/java/io/grpc/protobuf/StatusProto.java:38: error: cannot access GeneratedMessageV3
          ProtoLiteUtils.metadataMarshaller(com.google.rpc.Status.getDefaultInstance()));
                                                                 ^
  class file for com.google.protobuf.GeneratedMessageV3 not found
grpc-java/protobuf/src/main/java/io/grpc/protobuf/StatusProto.java:171: error: cannot access Builder
        .setCode(status.getCode().value());
        ^
  class file for com.google.protobuf.GeneratedMessageV3$Builder not found
2 errors
```

**Comment by ganadist** [comment]
https://protobuf.dev/news/2023-12-05

Also, I'm wondering how this protobuf changes will affect
* API changes
* Compatibility with artifacts which was generated with old protobuf/gRPC compilers.

**Comment by augi** [comment]
After updating to Protobuf 4.26.0, we observe this issue:
```
java.lang.NoClassDefFoundError: com/google/protobuf/GeneratedMessageV3
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
	at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(BuiltinClassLoader.java:862)
	at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(BuiltinClassLoader.java:760)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(BuiltinClassLoader.java:681)
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:639)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
	at io.grpc.health.v1.HealthGrpc.getCheckMethod(HealthGrpc.java:38)
	at io.grpc.health.v1.HealthGrpc.getServiceDescriptor(HealthGrpc.java:413)
	at io.grpc.health.v1.HealthGrpc.bindService(HealthGrpc.java:350)
	at io.grpc.health.v1.HealthGrpc$HealthImplBase.bindService(HealthGrpc.java:169)
```

**Comment by suztomo** [comment]
@neeme-praks-sympower Does any library or dependency require you to upgrade Protobuf runtime version to 4.x?

**Comment by be-hase** [comment]
It seems that `proto-google-common-protos`, which grpc-protobuf depends on, is not yet ver4 compliant.
https://github.com/googleapis/sdk-platform-java/tree/main/java-common-protos

**Comment by krickert** [comment]
Why couldn't grpc just bring back generated message v3?  The classes appear almost identical, I suppose it can just extend it?  

I tried that myself - it somewhat worked.  I got past the generated message v3 error, but I was getting serializaiton errors everywhere.  I think they may have changed the serializaiton spec too?

Anyway, I'm looking forward to this fix.  I might try this again today.

**Comment by neeme-praks-sympower** [comment]
> @neeme-praks-sympower Does any library or dependency require you to upgrade Protobuf runtime version to 4.x?

No. We just try to keep our dependencies up-to-date.

**Comment by suztomo** [comment]
@neeme-praks-sympower Thank you.

**Comment by ummels** [comment]
I've opened an [issue](https://github.com/protocolbuffers/protobuf/issues/16452) at Protobuf for restoring `GeneratedMessageV3` as an alias for `GeneatedMessage`.

**Comment by krickert** [comment]
> I've opened an [issue](https://github.com/protocolbuffers/protobuf/issues/16452) at Protobuf for restoring `GeneratedMessageV3` as an alias for `GeneatedMessage`.

I didn't think you're going to win that though

It was a major upgrade with breaking changes. They deprecated V3 awhile ago, better to fix instead of bringing it back. It'll take some time, but GRPC isn't always caught up to the latest protocol buffers.

**Comment by snazy** [comment]
The protobuf v4 release effectively breaks a ton of projects. Sure, it's a breaking change - but it is a breaking change _without_ a migration path. The protobuf code generators generated `GeneratedMessageV3` and that is now gone - so the tons of existing and working projects that use the protobuf code generators will stop working with protobuf v4.

IMHO, a breaking change _must_ first provide a properly communicated migration path and _then_ give all users and projects _enough_ time to adopt.

Bringing back GeneratedMessageV3 as a migration path SGTM - just not sure how good that'll work.

**Comment by ummels** [comment]
I'm in for a fight. :joy: 

Anyway, if they deprecated V3 a while ago, how come code generated with protoc `3.25.3` still relies on `GeneratedMessageV3`? Doesn't seem like a smooth transition to me. :thinking:

**Comment by ejona86** [comment]
Quick update: protobuf understands the problem and are considering solutions, and they've [said as much publicly](https://github.com/protocolbuffers/protobuf/issues/16452#issuecomment-2047931605). I expect a few more weeks to settle on the approach and then some time to make sure the details are right.

For those watching at home and want to be prepared, use protobuf-java 3.25 and protoc 25.  Also, if you have any non-generated code using `GeneratedMessageV3`, replace those usages with `Message`. Even if `GeneratedMessageV3` is added back to protobuf 4.x the newer generated code won't implement it. It was never intended to be referenced outside of generated code. I'm not aware of any reason to reference `GeneratedMessageV3` in hand-written code, so it is hopefully easy to avoid.

(I'm assuming there's no references to `GeneratedMessage` and protoc 2 since 3.0 came out in 2016. Any expected solution is expected to drop support for the 2.x generated code.)

**Comment by ejona86** [comment]
See https://github.com/protocolbuffers/protobuf/issues/16452#issuecomment-2076360676 for an update from protobuf. There's going to be a new 3.x release that will be compatible with 4.x. The ecosystem would upgrade to that new 3.x release before upgrading to 4.x. Once that release exists, gRPC will try to quickly pick it up, and might even backport the version bump (not something we typically do, but this is serious enough to probably be worth the risk).

**Comment by ejona86** [comment]
Protobuf 27.2 (protobuf-java 4.27.2) has restored compatibility with (at least some of) the 3.x generated code. https://github.com/protocolbuffers/protobuf/releases/tag/v27.2

grpc-java should now be compatible with protobuf-java 4.x. The Protobuf change was tested against 3.25.0, which means at least grpc-java 1.61+ should be compatible.

**Comment by krickert** [comment]
> Protobuf 27.2 (protobuf-java 4.27.2) has restored compatibility with (at least some later versions of) the 3.x generated code. https://github.com/protocolbuffers/protobuf/releases/tag/v27.2
> 
> grpc-java should now be compatible with protobuf-java 4.x. The Protobuf change was tested against 3.25.0, which means at least grpc-java 1.61+ should be compatible.

That's great news. I blame dependabot for making this bother me, but it wasn't that long of a wait at all.

**Comment by be-hase** [comment]
FYI: 
It seems that there are still some issues remaining with `protobuf-java`.

https://github.com/protocolbuffers/protobuf/issues/17247

**Comment by ejona86** [comment]
Appears the further issues may be resolved soon. Testing with the 4.28.0-RC3 looked good to me. https://github.com/protocolbuffers/protobuf/issues/17247#issuecomment-2305918282

**Comment by ejona86** [comment]
Protobuf 27.4 was released yesterday. We're targeting upgrading in 1.67.

**Comment by ascopes** [comment]
do we have a timescale for this, out of curiosity?

**Comment by ejona86** [comment]
@ascopes, "We're targeting upgrading in 1.67". The schedule for 1.67 is [in the milestone](https://github.com/grpc/grpc-java/milestone/79) (Sept 17th). But you don't have to wait on us. You can upgrade protobuf today in your own app.

**Comment by ascopes** [comment]
@ejona86 thanks, was asking as on the most recent version of protobuf and gRPC, the introspection service for gRPC-java was still failing due to the use of v3 APIs within protobuf.

**Comment by ascopes** [comment]
thanks, shall give it a try.

**Comment by ejona86** [comment]
The current plan is "wait and see." We're staying on 3.x until it seems our downstreams are successfully upgrading to 4.x. Basically, we're waiting until it is clear the ecosystem is compatible compatible with 4.x. That is most safely done from the leaves of dependency trees up to the root.

Notably, Google Cloud Java has a [separate BOM for protobuf java 3.x](https://cloud.google.com/java/docs/bom#updating_your_project_to_use_the_bom), which is identical to the regular BOM except which protobuf-java runtime version it uses. They're watching the uptake of the 4.x version and we're talking with a few larger companies to see if there's common libraries that need fixes before being compatible with 4.x.

Us staying on 3.x shouldn't impact anyone for now. But we do encourage using 4.x when able so that you're more prepared for protobuf-java 5.x. If one of your dependencies isn't compatible with 4.x, please file an issue with them (and if it is a commonly-used library/project, you can share a link to the issue here for our tracking).

**Comment by ejona86** [comment]
Summary of the situation for projects wondering what they should do:

 * Everyone: Replace all references of `GeneratedMessage` and `GeneratedMessageV3` outside of protoc-generated code with `Message` (or `MessageLite`, if you care about Android)
 * Applications: Try to upgrade to protobuf-java 4.27.4 or later 4.x version by adding an explicit dependency in your project. These versions are compatible with 3.x generated code using protoc 21.8 from Oct 2022. You can depend on protobuf-java 4.x even when your dependencies depend on protobuf-java 3.x. If you have problems with libraries being incompatible with 4.x, file bugs
 * Libraries:
   * For protobuf-java 3.x and 4.x support: Use protoc 25.x for your generated code, if at all possible. Versions as old as 21.8 seem to work, but haven't seen as much testing
   * For protobuf-java 4.x and 5.x support: drop protobuf-java 3.x support by upgrading to the latest protoc. You could upgrade to any protoc version after 26, which will all be compatible with protobuf-java 5.x. But latest is best
* Understand protobuf versioning:
  * protoc Y [is the same release](https://protobuf.dev/support/version-support/) as protobuf-java X.Y (the protobuf version becomes protobuf-java's minor version).
  * protoc Y generated code [requires at least](https://protobuf.dev/support/cross-version-runtime-guarantee/) protobuf-java X.Y. When running protoc yourself, there's no benefit to using an older protoc though; just match the versions between protoc and protobuf-java

**Comment by toddbaert** [comment]
I just wanted to confirm that for our library, @ejona86 's suggestion above worked. We are using protoc v25.5 (via the protobuf-maven-plugin and com.google.protobuf:protoc:3.25.5). We have set our minimum required protobuf-java version to `3.25.5`. Applications using 3.25+ of protobuf-java or 4+ of protobuf-java both work without any issues.

One thing we had to do was specifically pin the version of protobuf-java we use in our build (which generates sources) to 3.25.5 using a maven profile, otherwise the version range in our dependencies entry caused maven to resolve the latest 4.x version during the build, making it incompatible with applications using 3.x.

**Comment by jitrapon** [comment]
I would like to second that @ejona86's suggestion to explicitly declare `com.google.protobuf:protobuf-java` to be the same version as the `com.google.protobuf:protoc` works for me in my Gradle project as well.

**Comment by infa-kvaibhav** [comment]
@ejona86 Still after using  protobuf-java 4.27.4 , seems like  makeExtensionsImmutable() is removed from GeneratedMessageV3 which is causing method not found with oci-java-sdk-addons-sasl:3.55.1. Same method is persent in protobuf-java 3.x.

**Comment by abasu0713** [comment]
I have spent over 4 days trying this out to no avail. And it's a bit frustrating to be honest. Pardon my candor. 
I have multiple server implementations.. Some written in Dart, some in Go, some in Python.. I generate stubs myself. all using latest protoc version 4.29.3. I use multiple clients as well. Some written in Flutter and some in JS Frameworks. They all work seamlessly even with grpc-web with json transcoding for rest over Envoy. 
I need to write a Java based Events Listener SPI for Keycloak where I am using grpc java client to talk to dart and go servers as a post auth hook on keycloak. I have not been able to get it working. I keep getting the following error. 
```bash
2025-07-05 22:14:36,401 SEVERE [SignupEventListenerProvider] (executor-thread-3) Failed to create user using custom SPI: UNKNOWN: Channel Pipeline: [ProtocolNegotiators$PlaintextHandler#0, WriteBufferingAndExceptionHandler#0, DefaultChannelPipeline$TailContext#0]
``` 

My Maven dependencies look like the following:
```xml
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">

  <modelVersion>4.0.0</modelVersion>
  <groupId>com.example.auth</groupId>
  <artifactId>signup-spi</artifactId>
  <version>1.0.5</version>
  <packaging>jar</packaging>
  <name>custom-signup-spi</name>

  <build>
    <finalName>signup-spi</finalName> <!-- optional clean output name -->
    <extensions>
      <extension>
        <groupId>kr.motd.maven</groupId>
        <artifactId>os-maven-plugin</artifactId>
        <version>1.7.1</version>
      </extension>
    </extensions>
    <plugins>
      <!-- Protobuf Codegen -->
      <plugin>
        <groupId>org.xolstice.maven.plugins</groupId>
        <artifactId>protobuf-maven-plugin</artifactId>
        <version>0.6.1</version>
        <configuration>
          <protocArtifact>com.google.protobuf:protoc:3.25.5:exe:${os.detected.classifier}</protocArtifact>
          <pluginId>grpc-java</pluginId>
          <pluginArtifact>io.grpc:protoc-gen-grpc-java:1.73.0:exe:${os.detected.classifier}</pluginArtifact>
        </configuration>
        <executions>
          <execution>
            <goals>
              <goal>compile</goal>
              <goal>compile-custom</goal>
            </goals>
          </execution>
        </executions>
      </plugin>

      <!-- Shade Plugin for Guava conflict resolution -->
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-shade-plugin</artifactId>
        <version>3.5.1</version>
        <executions>
          <execution>
            <phase>package</phase>
            <goals>
              <goal>shade</goal>
            </goals>
            <configuration>
              <createDependencyReducedPom>false</createDependencyReducedPom>
              <relocations>
                <relocation>
                  <pattern>com.google.common</pattern>
                  <shadedPattern>com.biterush.shaded.guava</shadedPattern>
                </relocation>
              </relocations>
            </configuration>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>

  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>org.keycloak</groupId>
        <artifactId>keycloak-parent</artifactId>
        <version>26.2.5</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>

  <dependencies>
    <!-- Keycloak SPI Dependencies -->
    <dependency>
      <groupId>org.keycloak</groupId>
      <artifactId>keycloak-server-spi</artifactId>
      <version>26.2.5</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.keycloak</groupId>
      <artifactId>keycloak-server-spi-private</artifactId>
      <version>26.2.5</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.keycloak</groupId>
      <artifactId>keycloak-core</artifactId>
      <version>26.2.5</version>
    </dependency>

    <!-- gRPC Dependencies -->
    <dependency>
      <groupId>io.grpc</groupId>
      <artifactId>grpc-netty-shaded</artifactId>
      <version>1.63.0</version>
    </dependency>
    <dependency>
      <groupId>io.grpc</groupId>
      <artifactId>grpc-protobuf</artifactId>
      <version>1.63.0</version>
    </dependency>
    <dependency>
      <groupId>io.grpc</groupId>
      <artifactId>grpc-stub</artifactId>
      <version>1.63.0</version>
    </dependency>

    <!-- Protobuf Runtime -->
    <dependency>
      <groupId>com.google.protobuf</groupId>
      <artifactId>protobuf-java</artifactId>
      <version>3.25.3</version> <!-- doesn't work with 4.29.3 either -->
    </dependency>

    <!-- Guava (for gRPC internals) -->
    <dependency>
      <groupId>com.google.guava</groupId>
      <artifactId>guava</artifactId>
      <version>32.1.3-jre</version>
    </dependency>

    <!-- Java 17 compatibility (Keycloak requirement) -->
    <dependency>
      <groupId>org.apache.tomcat</groupId>
      <artifactId>annotations-api</artifactId>
      <version>6.0.53</version>
      <scope>provided</scope>
    </dependency>
  </dependencies>
</project>
```

Can someone give me some idea what is going on?

**Comment by ascopes** [comment]
@abasu0713 this probably isn't the answer but just if you were not aware, the newest version of protoc and protobuf-java is 4.31.1; I've got that working with grpc-java 1.73.0 in [basic integration tests](https://github.com/ascopes/protobuf-maven-plugin/tree/main/protobuf-maven-plugin/src/it/grpc-plugin) (nothing as complex as what you need though!) so it may be worth trying with that as well. I _doubt_ it will fix your issue but it rules that out at the very least.

**Comment by h-vetinari** [comment]
> I generate stubs myself. all using latest protoc version 4.29.3.

Just to note that latest right now is v31.1 (the baseline protobuf version; the leading major versions are different per language, i.e. 4.31.1 on the java side)

Edit: sorry, I hadn't seen that there was another comment to the same effect before I wrote my own.

**Comment by ejona86** [comment]
@abasu0713, I don't see anything about your report that makes it seem related to protobuf. Please open a new issue to discuss instead of doing it here.

Is there no more log information, like an exception stack trace or the like? Did you write the code for SignupEventListenerProvider?

**Comment by ejona86** [comment]
FYI: Protobuf 32 is adding support for older generated code. See the rc2 release notes:
https://github.com/protocolbuffers/protobuf/releases/tag/v32.0-rc2

**Comment by krickert** [comment]
Wanted to add to this thread.. I just added 4.x to my project and enforced it and it worked on the first shot!!! It's totally working and rockin'.. Been a long time comin.. I still have a lot of tests to do so YMMV

**Comment by ejona86** [comment]
Protobuf 4.33.0 supports all protobuf 3.x generated code: https://github.com/protocolbuffers/protobuf/releases/tag/v33.0

Do read the release notes, as it mentions how it handles vulnerabilities that exist in older generated code. Note that you still should not use `GeneratedMessage` or `GeneratedMessageV3` in your code; use `Message` instead.

I'm discussing with others when to upgrade. It seems it might happen soon. The largest source of remaining breakage is probably people using GeneratedMessageV3.
